File: algorithms/2_medium/O/00752_open_the_lock/bfs_hashset.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        invalids = set(deadends)
        q = deque()
        first_str = "0000"

        if first_str in invalids:
            return -1

        q.append(first_str)
        invalids.add(first_str)
        turns = 0

        while q:
            q_len = len(q)
            for _ in range(q_len):
                curr = q.popleft()

                if curr == target:
                    return turns

                curr_list = list(curr)
                temp_list = curr_list.copy()

                for i in range(len(curr_list)):
                    if curr_list[i] == '9':
                        temp_list[i] = '0'
                    else:
                        temp_list[i] = chr(ord(curr_list[i]) + 1)

                    temp = "".join(temp_list)

                    if not temp in invalids:
                        invalids.add(temp)
                        q.append(temp)

                    if curr_list[i] == '0':
                        temp_list[i] = '9'
                    else:
                        temp_list[i] = chr(ord(curr_list[i]) - 1)

                    temp = "".join(temp_list)

                    if not temp in invalids:
                        invalids.add(temp)
                        q.append(temp)

                    temp_list[i] = curr_list[i]
            turns += 1

        return -1

File: algorithms/2_medium/O/00752_open_the_lock/bfs_hashset.js
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
const modifyChar = (originalChar, changeAmount) => {
    const asciiValue = originalChar.charCodeAt(0);
    const modifiedAscii = asciiValue + changeAmount;
    return String.fromCharCode(modifiedAscii);
}
const openLock = (deadends, target) => {
    let invalids = new Set(deadends);
    let q = [];
    let first_str = "0000";

    if (invalids.has(first_str)) {
        return -1;
    }

    q.push(first_str);
    let turns = 0;

    while (q.length != 0) {
        const len = q.length;
        for (let c = 0; c < len; c += 1) {
            let curr = q.shift();

            if (curr == target) {
                return turns;
            }

            for (let i = 0; i < curr.length; i += 1) {
                const temp1 = curr.slice(0, i)  + (curr[i] === '9' ? '0' : modifyChar(curr[i], 1)) + curr.slice(i + 1);

                if (!invalids.has(temp1)) {
                    q.push(temp1);
                    invalids.add(temp1);
                }

                const temp2 = curr.slice(0, i)  + (curr[i] === '0' ? '9' : modifyChar(curr[i], -1)) + curr.slice(i + 1);

                if (!invalids.has(temp2)) {
                    q.push(temp2);
                    invalids.add(temp2);
                }
            }
        }

        turns += 1;
    }

    return -1;
};
